Имеются следующие данные о транзакциях в период с 01.12.2010 по 12.09.2011:

* InvoiceNo — номер транзакции
* StockCode — код товара
* Description — описание товара
* Quantity — количество единиц товара, добавленных в заказ
* InvoiceDate — дата транзакции 
* UnitPrice — цена за единицу товара
* CustomerID — id клиента
* Country — страна, где проживает клиент
* Данные можно скачать отсюда.

В данной задаче ссылка для считывания данных лежит в переменной path_to_file.
Импортируйте пандас и прочитайте данные с кодировкой ISO-8859-1. Запишите полученный датафрейм в retail, а названия колонок сохраните в переменную retail_columns.

In [2]:
import pandas as pd
retail = pd.read_csv("5_data.csv.zip",compression='zip', encoding='ISO-8859-1')
retail_columns = ['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country']
retail

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID         Country  
0        12/1/2010 8:26       2.55     17850.0  United Kingdom  
1        12/1/2010 8:26       3.39     17850.0  United Kingdom  
2        12/1/2010 8:26       2.75     17850.0  United Kingdom  
3        12/1/2010 8:26       3.39     17850.0  United Kingdom  
4        12/1/2010 8:26       3.39     17850.0  United Kingdom  
...                 ...        ...         ...             ...  
541904  12/9/2011 12:50       0.85     12680.0          France  
541905  12/9/2011 12:50       2.10     12680.0          France  
541906  12/9/2011 12:50       4.15     12680.0          France  
541907  12/9/2011 12:50       4.15     12680.0          France  
541908  12/9/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

Проверьте, встречаются ли в данных повторяющиеся наблюдения, и в качестве ответа укажите их количество. Если они есть, то удалите их из retail.


In [3]:
retail.duplicated().sum()

5268

In [4]:
retail = retail.drop_duplicates()
retail.shape[0]

536641

⭐️ Задание со звёздочкой! ⭐️

Данные содержат в себе записи как об успешных транзакциях, так и об отмененных. Если пользователь отменил заказ, в начале номера транзакции (InvoiceNo) ставится C (canceled). 

Сколько всего транзакций отменили пользователи? Считайте, что каждая строка - это отдельная транзакция.



In [7]:
retail.InvoiceNo.str.startswith("C").sum()

9251

Теперь отфильтруйте данные и оставьте в retail только те заказы, где Quantity > 0. В качестве ответа укажите число оставшихся строк.

In [30]:
filter_retail = retail.query("Quantity > 0")
filter_retail.shape[0]

526054

💀 Сложное задание! 💀

Посчитайте число заказов для каждого пользователя (CustomerID) из Германии (Germany). Оставьте только тех, кто совершил более N транзакций (InvoiceNo), где N – 80% процентиль. Запишите полученные id пользователей в germany_top (не весь датафрейм, только id).

In [34]:
retail

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID         Country  
0        12/1/2010 8:26       2.55     17850.0  United Kingdom  
1        12/1/2010 8:26       3.39     17850.0  United Kingdom  
2        12/1/2010 8:26       2.75     17850.0  United Kingdom  
3        12/1/2010 8:26       3.39     17850.0  United Kingdom  
4        12/1/2010 8:26       3.39     17850.0  United Kingdom  
...                 ...        ...         ...             ...  
541904  12/9/2011 12:50       0.85     12680.0          France  
541905  12/9/2011 12:50       2.10     12680.0          France  
541906  12/9/2011 12:50       4.15     12680.0          France  
541907  12/9/2011 12:50       4.15     12680.0          France  
541908  12/9/2011 12:50       4.95     12680.0          France  

[536641 rows x 8 columns]

In [21]:
germany_top = retail.query("Country == 'Germany'")\
    .groupby("CustomerID", as_index=False)\
    .agg({"InvoiceNo": "nunique"})\
    .query("InvoiceNo > InvoiceNo.quantile(q=0.80)")\
    .CustomerID


⭐️ Задание со звёздочкой! ⭐️

Используя объект с id пользователей (germany_top), полученный на предыдущем шаге, отфильтруйте наблюдения и оставьте в данных записи только по интересующим нас юзерам. Результирующий датафрейм запишите в top_retail_germany.

In [61]:
top_retail_germany =retail[retail.CustomerID.isin(germany_top)]
top_retail_germany

InvoiceNo StockCode                          Description  Quantity  \
1109      536527     22809              SET OF 6 T-LIGHTS SANTA         6   
1110      536527     84347  ROTATING SILVER ANGELS T-LIGHT HLDR         6   
1111      536527     84945   MULTI COLOUR SILVER T-LIGHT HOLDER        12   
1112      536527     22242        5 HOOK HANGER MAGIC TOADSTOOL        12   
1113      536527     22244           3 HOOK HANGER MAGIC GARDEN        12   
...          ...       ...                                  ...       ...   
541726    581570     22139     RETROSPOT TEA SET CERAMIC 11 PC          3   
541727    581570     23077                  DOUGHNUT LIP GLOSS         20   
541728    581570     20750             RED RETROSPOT MINI CASES         2   
541729    581570     22505            MEMO BOARD COTTAGE DESIGN         4   
541730    581570      POST                              POSTAGE         1   

            InvoiceDate  UnitPrice  CustomerID  Country  
1109    12/1/2010 13:04       2.95     12662.0  Germany  
1110    12/1/2010 13:04       2.55     12662.0  Germany  
1111    12/1/2010 13:04       0.85     12662.0  Germany  
1112    12/1/2010 13:04       1.65     12662.0  Germany  
1113    12/1/2010 13:04       1.95     12662.0  Germany  
...                 ...        ...         ...      ...  
541726  12/9/2011 11:59       4.95     12662.0  Germany  
541727  12/9/2011 11:59       1.25     12662.0  Germany  
541728  12/9/2011 11:59       7.95     12662.0  Germany  
541729  12/9/2011 11:59       4.95     12662.0  Germany  
541730  12/9/2011 11:59      18.00     12662.0  Germany  

[5540 rows x 8 columns]

Сгруппируйте top_retail_germany по коду товара (StockCode). Какой из продуктов добавляли в корзину чаще всего, кроме POST? 

Note: одним заказом считается единовременная покупка любого количества товара, т.е. без учета Quantity.


In [69]:
top_retail_germany.groupby("StockCode", as_index=False)\
    .Description\
    .count()\
    .query("StockCode !='POST'")\
    .sort_values(by="Description",ascending=False)\
    .head(1)

StockCode  Description
464     22326           75

Вернемся к анализу полного датасета retail. Создайте колонку Revenue с суммой покупки, используя колонки Quantity и UnitPrice.

In [70]:
retail["Revenue"] = retail.Quantity * retail.UnitPrice

In [71]:
retail

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID         Country  Revenue  
0        12/1/2010 8:26       2.55     17850.0  United Kingdom    15.30  
1        12/1/2010 8:26       3.39     17850.0  United Kingdom    20.34  
2        12/1/2010 8:26       2.75     17850.0  United Kingdom    22.00  
3        12/1/2010 8:26       3.39     17850.0  United Kingdom    20.34  
4        12/1/2010 8:26       3.39     17850.0  United Kingdom    20.34  
...                 ...        ...         ...             ...      ...  
541904  12/9/2011 12:50       0.85     12680.0          France    10.20  
541905  12/9/2011 12:50       2.10     12680.0          France    12.60  
541906  12/9/2011 12:50       4.15     12680.0          France    16.60  
541907  12/9/2011 12:50       4.15     12680.0          France    16.60  
541908  12/9/2011 12:50       4.95     12680.0          France    14.85  

[536641 rows x 9 columns]

Для каждой транзакции (InvoiceNo), посчитайте финальную сумму заказа. В качестве ответа укажите топ-5 (InvoiceNo) по сумме заказа (через запятую с пробелом и в порядке убывания TotalRevenue)

Например, для следующего примера...

`InvoiceNo StockCode Quantity InvoiceDate          UnitPrice  CustomerID  Revenue
536365    85123A    6        2010-12-01 08:26:00  2.55       17850.0     15.30
536365    71053     6        2010-12-01 08:26:00  3.39       17850.0     20.34
555555    71053     1        2010-12-02 18:00:00  3.39       17850.0     3.39`

...сумма заказа будет равна:

`InvoiceNo  TotalRevenue
536365     35.64
555555     3.39`

И ответом будет:

`536365, 555555`



In [78]:
top_revenue = retail.groupby("InvoiceNo", as_index=False)\
    .Revenue.sum()\
    .sort_values(by="Revenue",ascending=False)\
    .head(5)\
    .InvoiceNo
top_InvoiceNo = []
for i in top_revenue:
    top_InvoiceNo.append(i)
top_InvoiceNo

['581483', '541431', '574941', '576365', '556444']